json/decode.go

```diff
@@ -723,16 +736,24 @@ func (d *decodeState) literal(v reflect.Value) {
 	d.literalStore(d.data[start:d.off], v, false)
 }
 
-// convertNumber converts the number literal s to a float64 or a Number
-// depending on the setting of d.useNumber.
+// convertNumber converts the number literal s to a float64, int64 or a Number
+// depending on d.numberDecodeType.
 func (d *decodeState) convertNumber(s string) (interface{}, error) {
-	if d.useNumber {
+	if d.numberType == UnmarshalJSONNumber {
 		return Number(s), nil
 	}
+
 	f, err := strconv.ParseFloat(s, 64)
 	if err != nil {
 		return nil, &UnmarshalTypeError{"number " + s, reflect.TypeOf(0.0), int64(d.off)}
 	}
+
+	// if it has no decimal value use int64
+	if d.numberType == UnmarshalIntOrFloat {
+		if fi, fd := math.Modf(f); fd == 0.0 {
+			return int64(fi), nil
```

Comment:
If we use fi here rather than re-parsing from the string, won't we possibly lose precision? We'd cap out at 9,007,199,254,740,992 vs. the 18,446,744,073,709,551,615 that a uint64 could represent.

Reply:
good catch, the maximum safe integer in float representation is `2^53` but the maximum `int64` is 2^63-1
btw. you mentioned `uint64` but this PR is not meant for supporting unsigned integers but `int64`.

Reply:
fixed